instrumentation/jdbc/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/jdbc/test/JdbcInstrumentationTest.java

```diff
@@ -0,0 +1,1387 @@
+  }
+
+  @FunctionalInterface
+  public interface ThrowingBiConsumer<T, U> {
+    void accept(T t, U u) throws Exception;
+  }
+
+  static Stream<Arguments> recursiveStatementsStream() {
+    return Stream.of(
+        Arguments.of(
+            "getMetaData() uses Statement, test Statement",
+            false,
+            (ThrowingBiConsumer<Connection, String>)
+                (con, query) -> con.createStatement().executeQuery(query)),
+        Arguments.of(
+            "getMetaData() uses PreparedStatement, test Statement",
+            true,
+            (ThrowingBiConsumer<Connection, String>)
+                (con, query) -> con.createStatement().executeQuery(query)),
+        Arguments.of(
+            "getMetaData() uses Statement, test PreparedStatement",
+            false,
+            (ThrowingBiConsumer<Connection, String>)
+                (con, query) -> con.prepareStatement(query).executeQuery()),
+        Arguments.of(
+            "getMetaData() uses PreparedStatement, test PreparedStatement",
+            true,
+            (ThrowingBiConsumer<Connection, String>)
+                (con, query) -> con.prepareStatement(query).executeQuery()));
+  }
+
+  // regression test for
+  // https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/2644
+  @ParameterizedTest
+  @DisplayName("should handle recursive Statements inside Connection.getMetaData(): #desc")
+  @MethodSource("recursiveStatementsStream")
+  void testHandleRecursiveStatements(
+      String desc,
+      boolean usePreparedStatementInConnection,
+      ThrowingBiConsumer<Connection, String> executeQueryFunction)
+      throws Exception {
+    DbCallingConnection connection = new DbCallingConnection(usePreparedStatementInConnection);
+    connection.setUrl("jdbc:testdb://localhost");
+
+    testing.runWithSpan(
+        "parent",
+        () -> {
+          executeQueryFunction.accept(connection, "SELECT * FROM table");
+        });
+
+    testing.waitAndAssertTraces(
+        trace ->
+            trace.hasSpansSatisfyingExactly(
+                span -> span.hasName("parent").hasKind(SpanKind.INTERNAL).hasNoParent(),
+                span ->
+                    span.hasName("SELECT table")
+                        .hasKind(SpanKind.CLIENT)
+                        .hasParent(trace.getSpan(0))
+                        .hasAttributesSatisfying(
+                            equalTo(DbIncubatingAttributes.DB_SYSTEM, "other_sql"),
+                            equalTo(getDbConnectionStringKey(), "testdb://localhost"),
+                            equalTo(DbIncubatingAttributes.DB_STATEMENT, "SELECT * FROM table"),
+                            equalTo(DbIncubatingAttributes.DB_OPERATION, "SELECT"),
+                            equalTo(DbIncubatingAttributes.DB_SQL_TABLE, "table"),
+                            equalTo(ServerAttributes.SERVER_ADDRESS, "localhost"))));
+  }
+
+  static class DbCallingConnection extends TestConnection {
```

Comment:
I know this is where this class exists in the groovy file, but nested classes are typically put at the [bottom of the file](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/style-guideline.md#ordering-of-class-contents)